doc/aaa/radius_authentication.md

```diff
@@ -0,0 +1,634 @@
+   RADIUS plugin uses the cached MPL information to return user information.
+   If no MPL is available, and the many_to_one=a is configured, a least
+   privileged (MPL 1) temporary user info is returned. Otherwise, the
+   login fails.
+4. The PAM configuration causes a Access-Request to RADIUS server.
+5. RADIUS returns Access-Accept with Management-Privilege-Level (MPL)(Goto 6.),
+   or Access-Reject (Authentication Failed).
+6. Access-Accept: The MPL is cached.
+7. The user has authenticated successfully.
+
+## PAM
+
+The PAM to RADIUS authentication module allows any Linux device to become
+a RADIUS client for authentication. [pam_radius](#pam_radius)
+
+It supports many options in its configuration file including:
+
+- Multiple servers, with configurable ports, shared secrets, timeouts, VRF,
+  and source IP.
+- The order of the servers determines the priority.
+- Management-Privilege-Level attribute, which can be used to grant different
+  management access level to the users. This attribute is described in
+  Section 6.4 of [RFC5607](#RFC5607).
+
+This feature enhances this support with a source patch for the following:
+- CHAP authtype support.
+- PEAP/TLS MSCHAPv2 support. (with the use of freeradius library)
+
+This feature enhances this support with an external command that can be run
+with support from existing PAM modules.
+- Cache the Management-Privilege-Level attribute in a protected subdirectory
+  under /var/cache/radius/.
+
+This feature enhances this support with options supplied to existing PAM
+modules.
+- fail through mode of authentication.
+- The root user authentication should be only done against local.
+
+## NSS
+
+RADIUS (or any PAM authenticated) users are expected to have a configuration
+on the SONiC device to obtain the user information like uid, gid , home
+directory, and shell (i.e. information normally returned during a
+getpwnam_r() system library call). If the RADIUS user is found in the local
+password file, that information is used. Normally, RADIUS is only used
+for AAA, but not for nameservice. If another nameservice like LDAP is
+not used, a RADIUS NSS plugin needs to be used.
+
+The nss_radius plugin will load the configuration file /etc/radius_nss.conf.
+It will also use the Management-Privilege-Level (MPL)  attribute (type 136)
+information obtained from the RADIUS server during authentication, to
+determine the user information that is needed to create a local user, at
+the time of first login. The information of this local user is returned
+during a getpwnam_r() library call.
+
+There is an option to map all users at a privilege level to one user as
+specified in the configuration file /etc/radius_nss.conf. This option is
+the many_to_one=y option. When this option is in effect, no local user is
+created at the time of login. Because of the mapping, not all features
+may apply, since the LOGNAME of the user, and the mapped uid may not match.
```

Comment:
1) Are we planning to use this feature ?
2) If yes, how
3) How does this affects home dir and files created by user, his uid/gid ...?

Reply:
1. There is a corresponding (many_to_one=y)  option for TACACS+. I believe it would be used similarly. 
2. There are less use cases for this. It could be used, for example, if there is a requirement of creating only a fixed number of users (say one per privilege-level) on the SONiC device.
3. The user logging in will get the identity of the mapped user (Eg: If Jane's MPL was 15, and John's was 1, and the default table from section "User Privilege Table" were used, Jane would be operating as admin (uid=1000), and John would be operating as nobody (uid=65534). Thus all their files creations would be given the mapped identity of admin, and nobody. Similarly their home directories, by default without any other configuration, would be /home/admin and /nonexistent.)

Reply:
Added additional description for many_to_one=y case under NSS --> Mapping Users sub-section